File: 142_Linked_List_Cycle_II/142_3.java
public class Solution {
    public ListNode detectCycle(ListNode head) 
    {
        ListNode slow = head;
        ListNode fast = head;

        while (true)
        {
            if (fast == null || fast.next == null) return null;
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) break;
        }
        slow = head;
        while (slow != fast)
        {
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}

File: 142_Linked_List_Cycle_II/142_3.py
class Solution:
    def detectCycle(self, head):
        fast, slow = head, head
        while True:
            if not fast or not fast.next:
                return None
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                break

        slow = head
        while slow != fast:
            fast = fast.next
            slow = slow.next
        return slow